test/test_codegen.jl

```diff
@@ -210,87 +226,493 @@ end
+        @assert p.sorted_definitions == ["A", "B.G", "B"]
+        @test generate_struct_str(p.definitions["A"], ctx, remaining=Set{String}(["B.G", "B", "A"])) == """
+        struct var"##Stub#A"{T1<:var"##Abstract#B"} <: var"##Abstract#A"
+            b::Union{Nothing,T1}
         end
+        const A = var"##Stub#A"{var"##Stub#B"}
         """
         @test CodeGenerators.jl_default_value(p.definitions["A"].fields[1], ctx) == "nothing"
         @test CodeGenerators.jl_init_value(p.definitions["A"].fields[1], ctx) == "Ref{Union{Nothing,B}}(nothing)"
+
+        @test generate_struct_str(p.definitions["B.G"], ctx, remaining=Set{String}(["B.G", "B"])) == """
+        struct var"##Stub#B.G"{T1<:var"##Abstract#B"} <: var"##Abstract#B.G"
+            a::Union{Nothing,var"##Stub#A"{T1}}
+        end
+        const var"B.G" = var"##Stub#B.G"{var"##Stub#B"}
+        """
+        @test CodeGenerators.jl_default_value(p.definitions["B.G"].fields[1], ctx) == "nothing"
+        @test CodeGenerators.jl_init_value(p.definitions["B.G"].fields[1], ctx) == "Ref{Union{Nothing,A}}(nothing)"
+
+        @test generate_struct_str(p.definitions["B"], ctx, remaining=Set{String}(["B"])) == """
+        struct var"##Stub#B" <: var"##Abstract#B"
+            g::Union{Nothing,var"##Stub#B.G"{var"##Stub#B"}}
+        end
+        const B = var"##Stub#B"
+        """
+        @test CodeGenerators.jl_default_value(p.definitions["B"].fields[1], ctx) == "nothing"
+        @test CodeGenerators.jl_init_value(p.definitions["B"].fields[1], ctx) == "Ref{Union{Nothing,var\"B.G\"}}(nothing)"
+    end
+
+    @testset "Mutually recursive dependency involving a map field" begin
+        s, p, ctx = translate_simple_proto("syntax = \"proto3\"; message A { B b = 1; } message B { map<string, A> m = 1; }")
+        @assert p.sorted_definitions == ["A", "B"]
+        @test generate_struct_str(p.definitions["A"], ctx, remaining=Set{String}(["B", "A"])) == """
+        struct var"##Stub#A"{T1<:var"##Abstract#B"} <: var"##Abstract#A"
+            b::Union{Nothing,T1}
+        end
+        const A = var"##Stub#A"{var"##Stub#B"}
+        """
+        @test CodeGenerators.jl_default_value(p.definitions["A"].fields[1], ctx) == "nothing"
+        @test CodeGenerators.jl_init_value(p.definitions["A"].fields[1], ctx) == "Ref{Union{Nothing,B}}(nothing)"
+
+        @test generate_struct_str(p.definitions["B"], ctx, remaining=Set{String}(["B"])) == """
+        struct var"##Stub#B" <: var"##Abstract#B"
+            m::Dict{String,var"##Stub#A"{var"##Stub#B"}}
+        end
+        const B = var"##Stub#B"
+        """
+        @test CodeGenerators.jl_default_value(p.definitions["B"].fields[1], ctx) == "Dict{String,A}()"
+        @test CodeGenerators.jl_init_value(p.definitions["B"].fields[1], ctx) == "Dict{String,A}()"
+    end
+
+    @testset "Mutually recursive dependency involving where type param is shared with a different dependency" begin
```

Comment:
i don't think we have a test of recursive type with `parametrize_oneofs=true` _and_ `common_abstract_type=true`

Reply:
I've added a test case for that, thanks!